verifier/build/main.go

```diff
@@ -0,0 +1,345 @@
+			fmt.Sprintf("FT_LOG_URL=%s", *logURL),
+			fmt.Sprintf("FT_BIN_URL=%s", *binURL),
+			fmt.Sprintf("LOG_ORIGIN=%s", *logOrigin),
+			fmt.Sprintf("LOG_PUBLIC_KEY=%s", logFile),
+			fmt.Sprintf("APPLET_PUBLIC_KEY=%s", appletFile),
+			fmt.Sprintf("OS_PUBLIC_KEY1=%s", os1File),
+			fmt.Sprintf("OS_PUBLIC_KEY2=%s", os2File),
+		},
+		cleanup: func() {
+			os.RemoveAll(dir)
+		},
+	}
+}
+
+// releaseSigStore stores the signatures used for verifying releases.
+// Having such an object avoids needing to pass lots of state as args, or parsing
+// keys in different places.
+type releaseSigStore struct {
+	osV1    note.Verifier
+	osV2    note.Verifier
+	appV    note.Verifier
+	allV    note.Verifiers
+	envs    []string
+	cleanup func()
+}
+
+// Monitor verifiably checks inclusion of all leaves in a range, and then passes the
+// parsed FirmwareRelease to a handler.
+type Monitor struct {
+	st        client.LogStateTracker
+	stateFile string
+	sigStore  releaseSigStore
+	handler   func(context.Context, uint64, ftlog.FirmwareRelease) error
+}
+
+// From checks the leaves from `start` up to the checkpoint from the state tracker.
+// Upon reaching the end of the leaves, the checkpoint is persisted in the state file.
+func (m *Monitor) From(ctx context.Context, start uint64) error {
+	fromCP := m.st.LatestConsistent
+	pb, err := client.NewProofBuilder(ctx, fromCP, m.st.Hasher.HashChildren, m.st.Fetcher)
+	if err != nil {
+		return fmt.Errorf("failed to construct proof builder: %v", err)
+	}
+	klog.Infof("Running Monitor.From (%d, %d]", start, fromCP.Size)
+	var resErr error
+	for i := start; i < fromCP.Size; i++ {
+		klog.V(1).Infof("Leaf index %d: fetching leaf", i)
+		rawLeaf, err := client.GetLeaf(ctx, m.st.Fetcher, i)
+		if err != nil {
+			return fmt.Errorf("failed to get leaf at index %d: %v", i, err)
+		}
+		klog.V(1).Infof("Leaf index %d: fetching and verifying inclusion proof", i)
+		hash := m.st.Hasher.HashLeaf(rawLeaf)
+		ip, err := pb.InclusionProof(ctx, i)
+		if err != nil {
+			return fmt.Errorf("failed to get inclusion proof for index %d: %v", i, err)
+		}
+
+		if err := proof.VerifyInclusion(m.st.Hasher, i, fromCP.Size, hash, ip, fromCP.Hash); err != nil {
+			return fmt.Errorf("VerifyInclusionProof() %d: %v", i, err)
+		}
+
+		klog.V(1).Infof("Leaf index %d: parsing", i)
+		releaseNote, err := note.Open([]byte(rawLeaf), m.sigStore.allV)
+		if err != nil {
+			if e, ok := err.(*note.UnverifiedNoteError); ok && len(e.Note.UnverifiedSigs) > 0 {
+				return fmt.Errorf("unknown signer %q for leaf at index %d: %v", e.Note.UnverifiedSigs[0].Name, i, err)
+			}
+			return fmt.Errorf("failed to open leaf note at index %d: %v", i, err)
+		}
+
+		var release ftlog.FirmwareRelease
+		if err := json.Unmarshal([]byte(releaseNote.Text), &release); err != nil {
+			return fmt.Errorf("failed to unmarshal release at index %d: %w", i, err)
+		}
+
+		assertSigners := func(n *note.Note, names ...note.Verifier) error {
```

Comment:
Worth pulling this out into a named func to keep the size of the containing func down?

Reply:
Done.